Adjust orders module for varying statuses and updated database fields

jimbrig opened the pull request:
See #47 and #37 as well as [adminportal commit](https://github.com/powwater/adminportal/commit/8f35898bfb136406b77e516df237fe48774f059b) with details on database adjustments for orders table.

jimbrig commented:
@merlinoa for this code - [lines 160-162 of 2_orders_module.R](https://github.com/powwater/vendor_dashboard/blob/afd22eeb02e768bc971cbfd07d202838446b6104/shiny_app/R/2_orders_module.R#L160):

```r
out$order_status[2] <- "In Progress"
out$order_status[3] <- "Pending"
out$order_status[4] <- "Cancelled"
```

note that it is just for testing the styling and not representative of actual data. We well have to create some mock data for orders that are not complete/delivered to showcase functionality completely.

merlinoa commented on shiny_app/R/2_orders_module.R:
```diff
@@ -212,6 +254,20 @@ orders_module <- function(input, output, session, vendor_info){
 
+  num_awaiting_response <- reactive({
+    hold <- orders() %>%
+      filter(vendor_response == "Pending" || is.na(vendor_response)) %>%
```
`||` is the wrong logical operator to use here.  It is only making the or comparison for the first row which messes up the calculation.  If the first row is "Pending" of NA, this calculation results in all orders being awaiting response.  Use `|` instead.

jimbrig replied:
oops rookie mistake, fixed now

merlinoa commented:
The vendor needs to be able to click on an order with an NA vendor response and accept it or reject it.  Please add this functionality into this PR.  After the vendor either accepts or rejects the order, update the status columns accordingly.  Also, if the vendor rejects the order, require them to fill in the vendor response text.

merlinoa reviewed (changes requested):
see my comments above in the PR

jimbrig reviewed (commented):
@merlinoa this is ready for review again.

merlinoa commented:
When I click on an order that was rejected by the vendor I get this error:

![Screen Shot 2021-03-25 at 12 49 53 PM](https://user-images.githubusercontent.com/4030554/112511421-c8358d00-8d68-11eb-8264-3d8a9f9ab53f.png)

merlinoa commented:
If the order is rejected by the vendor, set the other order status columns to something other than "Pending" as those status are no longer "Pending".  NA would work, or maybe "Rejected".

merlinoa commented:
When I click cancel when accepting an order, the order is still accepted.  

![Screen Shot 2021-03-25 at 1 06 35 PM](https://user-images.githubusercontent.com/4030554/112513800-1481cc80-8d6b-11eb-9346-ae747d4181e1.png)

It should not be accepted in this situation.

merlinoa commented on shiny_app/R/2_orders_module.R:
```diff
@@ -66,78 +122,562 @@ orders_module <- function(input, output, session, vendor_info){
+      filter(order_uid == input$order_id_to_accept)
+  })
+
+  observeEvent(order_to_accept(), {
```
You do not need to nest reactives here.  Move reactives inside of this observeEvent outside of the observe Event.  You can use a reactiveVal to trigger the next reactive.  I think this was one reason for the bug when you click cancel in the accept order modal.

merlinoa commented on shiny_app/R/2_orders_module.R:
```diff
@@ -66,78 +122,562 @@ orders_module <- function(input, output, session, vendor_info){
+        )
+      )
+      session$userData$orders_trigger(session$userData$orders_trigger() + 1)
+      reloadData(awaiting_orders_table_proxy)
```
remove this reloadData().  It is not necessary.  We only want to reload the data after the new data (i.e. the data with the newly accepted order) is fetched from the db.

merlinoa commented on shiny_app/R/2_orders_module.R:
```diff
@@ -66,78 +122,562 @@ orders_module <- function(input, output, session, vendor_info){
+        ) %>%
+        tibble::add_column(" " = actions, .before = 1)
+
+      addRow(orders_table_proxy, out, resetPaging = FALSE)
```
Remove everything from this lines 490 to this line.  We do not want to reload the table at this point.  We want to wait for the data from the database.

merlinoa commented on shiny_app/R/2_orders_module.R:
```diff
@@ -66,78 +122,562 @@ orders_module <- function(input, output, session, vendor_info){
+
+  orders_prep <- reactiveVal(NULL)
+
+  observeEvent(list(orders(), awaiting_orders()), {
```
Do not trigger this when awaiting_orders() changes.  Filter awaiting_orders() out of the orders() data frame before it gets to this reactive.  There is not reason that a change to awaiting_orders() should require a refresh of the order().  Only a change to orders() should change the orders table.

merlinoa commented:
"rejected_time" is not being populated in the database.  Change this column name to "vendor_response_time" and populate is when the order is either rejected or accepted.

merlinoa commented:
Do not put the user's phone number in the modified_by or created_by columns.  Use the users uid from `session$userData$user()user_uid`

merlinoa reviewed (changes requested):
Please see my comments above.  For the comments that do not have a "Resolve Conversation" button, can you add a thumbs up after you make the fix.  Thanks

jimbrig commented:
@merlinoa all changes have been incorporated into the orders module. Code is much easier now and app is looking good.

Note that you will need to update your version of [powpolished](https://github.com/powwater/powpolished) to be the latest version from GitHub, as well as get the new `config.yml` (this should be done automatically when you run the app though).

 On top of your comments, and various other changes I also added the following features: 

- New ratings widget at bottom of page (not sure where's best currently, separate tabset?):

![image](https://user-images.githubusercontent.com/32652297/112558657-9f7fb880-8da5-11eb-8ce0-65b44af69ee3.png)

- added utility functions for formatting columns and HTML escaping, etc.
- automated deployment via new dev functions and a Dockerfile template - see [this commit](https://github.com/powwater/vendor_dashboard/commit/ada3b4b2cfc8f026d69da93223ceee5067e8b371) for details:

```
library(polished)
library(yaml)
library(whisker)
library(usethis)

source("dev/sysreqs.R")
source("dev/pkgdeps.R")
source("dev/use_template.R")

# gather R package dependencies -------------------------------------------
deps <- polished:::get_package_deps("shiny_app")
yaml::write_yaml(deps, "shiny_app/deps.yml")

# gather sysreqs ----------------------------------------------------------
sysreqs <- get_sysreqs(names(deps))

# command strings ---------------------------------------------------------
cran_install_cmd <- get_cran_deps(deps) %>% cran_packages_cmd()
gh_install_cmd <- get_gh_deps(deps) %>% gh_packages_cmd()
sysreqs_cmd <- paste(paste0("RUN ", apt_get_install(sysreqs), collapse = " \\ \n"))

# create Dockerfile from template -----------------------------------------
use_template("dev/Dockerfile_template",
             "Dockerfile",
             data = list(
               sysreqs = sysreqs_cmd,
               cran_installs = cran_install_cmd,
               gh_installs = gh_install_cmd
             ))
```

creates:

```
FROM gcr.io/powwater/shiny_run_server

ARG R_CONFIG_ACTIVE=default

# System Requirements
RUN apt-get update -qq && apt-get -y --no-install-recommends install \ 
  libjpeg-dev \ 
  libcurl4-openssl-dev \ 
  libssl-dev \ 
  make \ 
  default-jdk \ 
  git \ 
  libicu-dev \ 
  pandoc \ 
  libpq-dev \ 
  libgit2-dev

RUN R -e "install.packages('remotes')"

# CRAN R packages 
RUN R -e "remotes::install_version('DBI', version = '1.1.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('RPostgres', version = '1.3.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('dbplyr', version = '2.1.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('dplyr', version = '1.0.5', upgrade = 'never')" 
RUN R -e "remotes::install_version('DT', version = '0.17', upgrade = 'never')" 
RUN R -e "remotes::install_version('formattable', version = '0.2.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('snakecase', version = '0.11.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('htmltools', version = '0.5.1.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('shiny', version = '1.6.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('shinycustomloader', version = '0.9.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('shinydashboard', version = '0.7.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('shinydashboardPlus', version = '0.7.5', upgrade = 'never')" 
RUN R -e "remotes::install_version('shinyjs', version = '2.0.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('shinyWidgets', version = '0.6.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('yaml', version = '2.2.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('lubridate', version = '1.7.10', upgrade = 'never')" 
RUN R -e "remotes::install_version('purrr', version = '0.3.4', upgrade = 'never')" 
RUN R -e "remotes::install_version('rlang', version = '0.4.10', upgrade = 'never')" 
RUN R -e "remotes::install_version('tidyselect', version = '1.1.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('attempt', version = '0.3.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('googleway', version = '2.7.3', upgrade = 'never')" 
RUN R -e "remotes::install_version('shinyFiles', version = '0.9.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('waiter', version = '0.2.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('dialr', version = '0.3.2', upgrade = 'never')" 
RUN R -e "remotes::install_version('config', version = '0.3.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('tibble', version = '3.1.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('stringr', version = '1.4.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('scales', version = '1.1.1', upgrade = 'never')" 
RUN R -e "remotes::install_version('fs', version = '1.5.0', upgrade = 'never')" 
RUN R -e "remotes::install_version('usethis', version = '2.0.1', upgrade = 'never')"

# GitHub R packages 
RUN R -e "remotes::install_github('powwater/powpolished', ref = '5e38b291fdb2103e8c8247db4811a2b54ea0d13c', upgrade='never')" 
RUN R -e "remotes::install_github('tychobra/tychobratools', ref = 'c1a24b413363174f1115a0b34d9ae3f8616e5b76', upgrade='never')" 
RUN R -e "remotes::install_github('merlinoa/shinyFeedback', ref = 'ceb65e02428181a166a8b2cb20aac727c6f261ab', upgrade='never')" 
RUN R -e "remotes::install_github('JohnCoene/shinyscroll', ref = '98f5d669ab76c9e44fda798beac4c17dd58f66c9', upgrade='never')"

COPY 'shiny_app' /srv/shiny-server/shiny_app

RUN chown -R shiny:shiny /srv/shiny-server/shiny_app

RUN echo "Sys.setenv(R_CONFIG_ACTIVE='$R_CONFIG_ACTIVE')" >> /home/shiny/.Rprofile

USER shiny

CMD ["/usr/bin/shiny-server"]
```

using the template approach.

jimbrig reviewed (commented):
@merlinoa updated with your comments

merlinoa commented:
I was supposed to review this before merge...
